selfdrive/car/car_helpers.py

```diff
@@ -48,48 +42,50 @@ def _get_interface_names() -> dict[str, list[str]]:
 interfaces = load_interfaces(interface_names)
 
 
-def can_fingerprint(next_can: Callable) -> tuple[str | None, dict[int, dict]]:
+def can_fingerprint(logcan: SimpleNamespace) -> tuple[str | None, dict[int, dict]]:
   finger = gen_empty_fingerprint()
   candidate_cars = {i: all_legacy_fingerprint_cars() for i in [0, 1]}  # attempt fingerprint on both bus 0 and 1
   frame = 0
   car_fingerprint = None
   done = False
 
   while not done:
-    a = next_can()
-
-    for can in a.can:
-      # The fingerprint dict is generated for all buses, this way the car interface
-      # can use it to detect a (valid) multipanda setup and initialize accordingly
-      if can.src < 128:
-        if can.src not in finger:
-          finger[can.src] = {}
-        finger[can.src][can.address] = len(can.dat)
-
+    # logcan.drain(wait_for_one=True) may return empty list or multiple packets,
```

Comment:
This would be rare, but on cars with hardly any CAN fingerprints, to the point where some packets return empty from boardd, the previous function would possibly wait longer than the 100 frames/1 second we intended, as `get_one_can` used to wait for a non-empty can packet. This now changed to be constant time in this rare case

Reply:
Though we can maintain previous behavior with: https://github.com/commaai/openpilot/pull/33215/commits/f848fd32132391692c6191a305bb38f74091ec91

Reply:
Did in this PR: https://github.com/commaai/openpilot/pull/33232